### ISSUE TITLE
Short-circuit behavior for the && and || operators

### ISSUE THREAD

**Issue by vburckhardt** [opentofu/opentofu#828]
### OpenTofu Version

```shell
All of them
```

### Use Cases

&& and || operators are not short-circuiting in terraform, unlike most modern language. This is a very common confusion - see https://github.com/hashicorp/terraform/issues/24128

This post seems to indicate that the issue is complicated to solve in tf due to an upstream dependency, which may not be an issue in opentufu. https://github.com/hashicorp/terraform/issues/24128#issuecomment-702456250

Changing the behavior was previously explored in details at https://github.com/hashicorp/hcl/pull/538

### Attempted Solutions

Various workaround exists for this general limitation - most commonly using `condition ? true_val : false_val`

### References

https://github.com/hashicorp/terraform/issues/24128
https://github.com/hashicorp/hcl/pull/538

**Comment by cube2222** [comment]
Thanks for submitting the issue @vburckhardt!

We do in fact use the same dependency and would prefer not to fork it unless absolutely necessary. However, I do think it would make sense to bump that PR or ask there, since there's definitely willingness to work on it in the HCL repo if a PR is opened.

**Comment by janosdebugs** [comment]
Hey @vburckhardt thank you for submitting this issue and sorry for the long delay. We have discussed this issue in the core team and were wondering what your use case for this issue is? Specifically, what is it that you can't do today that you would be able to do with this?

**Comment by apparentlymart** [comment]
There is a new upstream HCL issue for this at https://github.com/hashicorp/hcl/pull/713, which seems like it's headed toward being merged at the time I'm writing this comment.

**Comment by cam72cam** [comment]
Looks like there is at least one followup PR that we should keep an eye on

**Comment by joe-a-t** [comment]
I think this is unblocked now, can we get this added to the roadmap?

**Comment by cam72cam** [comment]
I'd still like to wait for a hcl release with these changes before adding it to the milestone

**Comment by apparentlymart** [comment]
As far as I can tell from the commit history and tags in the HCL repository, this functionality has not yet been included in a tagged HCL release. The most recent release is v2.23.0, but the commits related to the short-circuit behavior were added after that tag.

I can't say how this was already included in the other project, but the Go toolchain does support pointing to a specific commit rather than to a tag, so we could consider doing that. (We already have [our own fork of HCL](https://github.com/opentofu/hcl) anyway, so we're already using something different than one of the upstream tags and could potentially cherry-pick those changes into our fork.)

**Comment by cam72cam** [comment]
This has already been added to our fork a while ago during other changes.  Closing.
